libnetwork/ipam/allocator.go

```diff
@@ -425,17 +413,16 @@ func (a *Allocator) DumpDatabase() string {
 		for k, config := range aSpace.subnets {
 			s += fmt.Sprintf("\n%v: %v", k, config)
 			if config.Range == nil {
-				a.retrieveBitmask(k, config.Pool)
+				aSpace.retrieveBitmask(k, config.Pool)
```

Comment:
I think this may cause a deadlock here.
Since `aSpace` is already locked and `retrieveBitmask` also locks.

There's some other not great locking in addrSpace (e.g. `a.Lock(); doStuff; a.Unlock(); a.aThingThatLocksAgain()`)

Reply:
I completely agree, which is why I address all of that in later commits. This DumpDatabase function is completely unused; I'm not really sure why I'm bothering to maintain it.

Reply:
I've squashed the offending commits together